### PR TITLE
Resurrect auto-port reassignment, but for all providers

### PR DESCRIPTION
**Pull request by n1hility**
- Move directory location code to break import cycles
- Updates common to main to pull in edit config locking

```release-note
none
```

### CODE REVIEW

File: pkg/machine/env/dir.go
Comment (Luap99) [inline pkg/machine/env/dir.go:1]: Moving this to env feels kinda arbitrary and I do not really like it. Also makes reviewing this diff much harder.
I am not sure where exactly you import cycle happens.

File: pkg/machine/qemu/options_windows_amd64.go
Comment (Luap99) [inline pkg/machine/qemu/options_windows_amd64.go:1]: ?

File: pkg/machine/qemu/stubber.go
Comment (Luap99) [inline pkg/machine/qemu/stubber.go:1]: see https://github.com/containers/podman/pull/21686
AFAIK the goal is to have qemu build on windows there so we should not make it worse here without any reason.

**Review by Luap99 — COMMENTED** [review]
I rather not move so much stuff around if possible, anyhow will leave it to @baude @ashley-cui to decide where this code should move

**Comment by baude** [comment]
why do we have to move he machine code?

File: pkg/machine/env/dir.go
Comment (n1hility) [reply pkg/machine/env/dir.go:1]: The cycle I hit was machine -> connection -> lock -> machine. Since these routines are frequently accessed from many machine  packages the chances of a cycle are high (hit this one before), and since they have no natural dependency on any other package (just the system environment) it makes sense to move them somewhere where they can be a leaf node in the graph. Since they are all computing directories/paths from the env, it IMO makes sense organizationally as well.  Although I am not sold on the name, went pack in forth between paths/env/dirs.

File: pkg/machine/qemu/stubber.go
Comment (n1hility) [reply pkg/machine/qemu/stubber.go:1]: The problem is that the qemu provider does compile on windows && mac because of a purposeful change that stubs out core data types (e.g. pkg/machine/vmconfigs/config_windows.go), so since these files are broken with the existing tags they set off validation / lint errors. This is just a temporary least invasive workaround.

File: pkg/machine/env/dir.go
Comment (n1hility) [reply pkg/machine/env/dir.go:1]: (BTW in hindsight I should have split the refactor in a separate commit to make it easier to review, will try to do that more in the future)

File: pkg/machine/qemu/options_windows_amd64.go
Comment (n1hility) [reply pkg/machine/qemu/options_windows_amd64.go:1]: see other comment on qemu/stubber

**Comment by ashley-cui** [comment]
FWIW I don't have a particularly strong opinion on moving this machine code but am not convinced of the name of the new pkg.

File: pkg/machine/env/dir.go
Comment (Luap99) [reply pkg/machine/env/dir.go:1]: Yeah a separate commit would help a lot for this stuff, also including the import path.

So wouldn't it be easier to add the lock in c/common, I abstracted all the writing code into config.EditConnectionConfig() so locking there would only require 1 place which is much more desirable and safer as we cannot forget anything.

Comment (Luap99) [reply pkg/machine/env/dir.go:1]: How about I open a PR for c/common to do the locking there and then you can just leave the code where it was and only add the port logic?

**Comment by n1hility** [comment]
> why do we have to move he machine code?

I needed some of these env methods to be reachable from machine/connection (GetDataDir/GetRuntimeDir etc) , which introduced a cycle. Since they have no natural dependency (just calculating paths of env vars / system constants) the cleanest approach was  to move them into a leaf package of some form. Although not hung up on the name.

Some more details in my other reply here:  https://github.com/containers/podman/pull/21927#discussion_r1511233410

**Comment by Luap99** [comment]
I created https://github.com/containers/common/pull/1880 for the connections file lock.

File: pkg/machine/env/dir.go
Comment (n1hility) [reply pkg/machine/env/dir.go:1]: I agree the locking logic totally belongs in common for sure. Only reason I didn't put it there initially was just being up against the gun on 5.0 and trying to avoid a common pull, but certainly the way we need to do it (moving the locking to the single edit spot)

I do think the cycles on these functions will keep happening though, so at some point I think it will need to be refactored in this direction. Just a question of 5.0 or a later release.

Comment (Luap99) [reply pkg/machine/env/dir.go:1]: Ack I am not blocking over the move then. I can live with having it in env as I cannot offer a better package name for it.

**Comment by n1hility** [comment]
@containers/podman-maintainers PTAL, should be good to go, other than it pulls in common@main. I reworked the commits to split off the env refactor into a separate commit, to make it easier to review (amounts to the same total delta but you can click each commit to breakout apart). I was going to offer to defer/drop the env refactor to post 5.0, but  Its still necessary for the port change since there is another cycle between vmconfig and machine. @Luap99 and I discussed this a bit in detail and he seems cool with it now: https://github.com/containers/podman/pull/21927#discussion_r1511421605

**Comment by n1hility** [comment]
rebased after another go.mod conflict

**Review by Luap99 — APPROVED** [review]
LGTM
@baude @ashley-cui PTAL again

**Comment by mheon** [comment]
LGTM

**Comment by baude** [comment]
Im going to merge  but i wonder if any tests could be written for this? at least a regression test?  at least for windows?

/lgtm

**Comment by n1hility** [comment]
> Im going to merge but i wonder if any tests could be written for this? at least a regression test? at least for windows?
> 
> /lgtm

Yeah sorry I should have addressed that. I plan to follow up with a regression test as part 2. Just wanted to get things in before blowing up your feature freeze.

File: pkg/machine/shim/networking.go
Comment (siretart) [inline pkg/machine/shim/networking.go:14]: I believe this is what broke the compilation on pretty much every architecture but arm64 and amd64.

See https://buildd.debian.org/status/package.php?p=libpod&suite=experimental for build logs:

```
package github.com/containers/podman/pkg/machine/shim
	imports github.com/containers/podman/v5/pkg/machine/connection: build constraints exclude all Go files in /<<PKGBUILDDIR>>/_output/src/github.com/containers/podman/pkg/machine/connection
```

Is `pkg/machine/shim` supposed to be compiled on those architectures or not? 
cc: @n1hility  @mheon @baude @vrothberg

Comment (Luap99) [reply pkg/machine/shim/networking.go:14]: FYI: I suggest you create a new issues instead of commenting on old PRs.

That said podman compiles just fine on other arches as well, the issue how your debian build tooling works.  As machine is only compiled on amd64 and arm64 this package is not in the import chain and as such will not cause any compile errors. Your build chain is breaking because it calls something like go build ./pkg/machine/shim which will then of course fails on the other architectures as it is not designed to build there.

It is basically the same as https://github.com/containers/podman/issues/10625. So we can(should?!) add the build tags to the shim package as well to make it consistent with the rest.

**Pull request merged**
